### ISSUE TITLE
[Help] Last-Will not working

### ISSUE THREAD

**Issue by gabrielmancini** [emqx/MQTTX#899]
#### Describe the problem you Confuse

i set the Last-Will Topic, Qos, Retain, Payload, connect and disconnect and the broker not send the last will message,
i am using emqx on docker 

#### More detail (optional)

my connection is not save and i am using the port 1883

**Comment by ysfscream** [comment]
Hi, the last will message will not be sent simply by disconnecting, it will be triggered only if the following conditions are met. When the client disconnects, a will message is sent to the relevant subscriber. Will Messages will be sent when:

- I / O error or network failure occurred on the server;
- The client loses contact during a defined heartbeat period;
- The client closes the network connection before sending offline packets;
- The server closes the network connection before receiving the offline packet.

Thanks, ref: https://www.emqx.com/en/blog/use-of-mqtt-will-message

**Comment by gabrielmancini** [comment]
there is a way to test last will on my broker using MQTTX?

**Comment by ysfscream** [comment]
In order to achieve the effect that MQTT connections are abnormally disconnected, we need to adjust EMQX's default ACL rules and related configuration items.

First, add the following ACL rule to `etc/acl.conf` to deny local clients from connecting to the test topic. Note that this rule needs to be added before all the default ACL rules to ensure that it will work.

{deny, {ipaddr, "127.0.0.1"}, publish, ["test"]}.
Then modify the zone.internal.acl_deny_action configuration in etc/emqx.conf to set it to disconnect the client when the ACL check denies: {deny, {ipaddr, "127.0.0.1"}, publish, ["test"}.

zone.internal.acl_deny_action = disconnect
Once we have made these changes, we start EMQX.

Next, we create a new connection named demo in MQTT X, change the Host to localhost, select MQTT Version 5.0 in the Advanced section, and set the Session Expiry Interval to 10 to ensure that the session does not expire before the probate message is published.

![image](https://user-images.githubusercontent.com/21299158/157149536-0583508a-bed6-4467-bc60-c2a6b0cf5b41.png)

Then in the Last Will and Testament section set Last-Will Topic to offline, Last-Will Payload to I'm offline, and Will Delay Interval (s) to 5.

![image](https://user-images.githubusercontent.com/21299158/157149569-237cb9e9-78d7-49fa-bfd4-1b7be61b609e.png)

Once the above settings are complete, we click the Connect button in the upper right corner to establish the connection.

Then we create a client connection named subscriber and subscribe to the offline topic.

![image](https://user-images.githubusercontent.com/21299158/157149609-b8a2f1ba-8dfd-41fc-acdb-d846e008be7d.png)

Next, we go back to the demo connection and post a message with any content with Topic as a test, at this point, the connection will be disconnected, wait five seconds and we will see the subscriber connection receive a will message with the content `I'm offline`.

![image](https://user-images.githubusercontent.com/21299158/157149679-adf335f2-c202-4acb-86e7-6a10c7c00586.png)

**Comment by gabrielmancini** [comment]
awesome!

**Comment by gabrielmancini** [comment]
sorry i closed accidentally

**Comment by ysfscream** [comment]
> awesome! its just woks on version 5?

YES. You can select MQTT version 5.0 and test it again.

**Comment by gabrielmancini** [comment]
let me rephrase, Last will not work on 3.1?

**Comment by ysfscream** [comment]
> let me rephrase, Last will not work on 3.1?

Oh, sorry. I did not read your question carefully, of course not, the Last Will can be used in all versions
